### ISSUE TITLE
集群上event_hander打印测试集cost不对

### ISSUE THREAD

**Issue by prm10** [PaddlePaddle/Paddle#4633]
``` python
    # Define end batch and end pass event handler
    def event_handler(event):
        if isinstance(event, paddle.event.EndIteration):
            if event.batch_id % 100 == 0:
                print "\nPass %d, Batch %d, Cost %f, %s" % (
                    event.pass_id, event.batch_id, event.cost, event.metrics)
            #else:
            #    sys.stdout.write('.')
            #    sys.stdout.flush()
        if isinstance(event, paddle.event.EndPass):
            result = trainer.test(reader=test_reader, feeding=feeding)
            print "\nTest with Pass %d, %s" % (event.pass_id, result.metrics)
```
在event_hander中写了上面代码，在单机环境下，得到的测试集cost与每轮batch训练集打印的cost接近，每轮训练都会一起逐步下降，但是在mpi集群运行时，打印的结果如下：
![image](https://user-images.githubusercontent.com/14084414/31327367-6a8753e0-ac94-11e7-9905-f9bb17e9500e.png)
训练集的cost从最开始的0.7逐渐下降到0.3；在单机上，测试集的cost也会下降到0.3附近，但是集群上的测试集cost则一直没变过。
甚至当我把测试集的数据替换为与训练集一模一样之后，测试集的cost也没有丝毫下降。

**Comment by NHZlX** [comment]
单机以及集群的的batch/learning rate 参数各是多少呢

**Comment by prm10** [comment]
batch 都是200，learning rate 都是1e-3；集群node数20。

**Comment by typhoonzero** [comment]
可以试试集群的learning rate调整成 `1e-3 / 20`

**Comment by prm10** [comment]
刚试过了，测试集的cost仍然没有下降。
需要注意的是，我已经把测试集数据改为了训练集（二者完全相同了），所以理论上测试集的cost必然会下降，但是在集群训练中却没有变化。
所以想请你们确认下是不是个bug

**Comment by typhoonzero** [comment]
不应该，检查下代码，train部分需要指定is_local参数：
```python
trainer = paddle.trainer.SGD(cost,
                                 parameters,
                                 opt,
                                 is_local=false)
```

**Comment by prm10** [comment]
还是不行，这里贴出代码，求指正：
``` python
# -*- coding: utf-8 -*-
# Created by pangrenming@baidu.com on 2017/9/28
import sys, os
import paddle.v2 as paddle
import gzip
import random

# 集群配置
node_id = os.getenv("OMPI_COMM_WORLD_RANK")
if not node_id:
    raise EnvironmentError("must provied OMPI_COMM_WORLD_RANK")
TRUTH = ["true", "True", "TRUE", "1", "yes", "Yes", "YES"]
cluster_train = os.getenv('PADDLE_CLUSTER_TRAIN', "False") in TRUTH
use_gpu = os.getenv('PADDLE_USE_GPU', "False")
# 集群参数
cluster_train_dir = "./train_data_dir/train"
cluster_test_dir = "./test_data_dir/test"


def save_model(parameters, save_path):
    with open(save_path, 'w') as f:
        parameters.to_tar(f)


def load_dict(dict_filename, dict_size):
    word_dict = dict()
    with open(dict_filename) as f:
        i = 0
        for line in f:
            line = line.strip()
            word = line.split('\t')[0]
            word_dict.setdefault(word, i)
            i += 1
            if i == dict_size:
                break
    return word_dict


def parse_query(query, word_dict):
    ret = list()
    for word in query.split('|'):
        if word in word_dict:
            ret.append(word_dict[word])
    return ret


def parse_title(title, word_dict):
    ret = list()
    for word in title.split('|'):
        if word in word_dict:
            ret.append(word_dict[word])
    return ret


def parse_content_seq(content, word_dict):
    ret = list()
    for sentence in content.split('#'):
        for word in sentence.split('|'):
            if word in word_dict:
                ret.append(word_dict[word])
    return ret


def parse_content_hie(content, word_dict):
    ret = list()
    for sentence in content.split('#'):
        word_list = list()
        for word in sentence.split('|'):
            if word in word_dict:
                word_list.append(word_dict[word])
        if len(word_list) > 0:
            ret.append(word_list)
    return ret


def reader_pairwise_content(data_filedir, dict_filename, dict_size):
    def reader():
        word_dict = load_dict(dict_filename, dict_size)
        for filename in os.listdir(data_filedir):
            data_filename = os.path.join(data_filedir, filename)
            with open(data_filename) as f:
                for line in f:
                    line = line.strip()
                    info = line.split('\t')
                    if len(info) == 7:
                        q, t1, t2, c1, c2, u1, u2 = info
                        q_ret = parse_query(q, word_dict)
                        c1_ret = parse_content_hie(c1, word_dict)
                        c2_ret = parse_content_hie(c2, word_dict)
                        if len(q_ret) == 0 or len(c1_ret) == 0 or len(c2_ret) == 0:
                            continue
                        if random.randint(0, 1) == 1:
                            yield q_ret, c1_ret, c2_ret, 1
                        else:
                            yield q_ret, c2_ret, c1_ret, 0

    return reader


def reader_pairwise_title(data_filedir, dict_filename, dict_size):
    def reader():
        word_dict = load_dict(dict_filename, dict_size)
        for filename in os.listdir(data_filedir):
            data_filename = os.path.join(data_filedir, filename)
            with open(data_filename, 'r') as f:
                for line in f:
                    line = line.strip()
                    info = line.split('\t')
                    if len(info) == 7:
                        q, t1, t2, c1, c2, u1, u2 = info
                        q_ret = parse_query(q, word_dict)
                        t1_ret = parse_title(t1, word_dict)
                        t2_ret = parse_title(t2, word_dict)
                        if len(q_ret) == 0 or len(t1_ret) == 0 or len(t2_ret) == 0:
                            continue
                        if random.randint(0, 1) == 1:
                            yield q_ret, t1_ret, t2_ret, 1
                        else:
                            yield q_ret, t2_ret, t1_ret, 0

    return reader


def get_sum_model(scope_name, seq_data, word_vector_dim, fc_dim, fc_param):
    embedding_param = paddle.attr.Param(
        name='embedding_param',
        initial_max=1.0,
        initial_min=-1.0,
        initial_std=0.01,
    )
    fc_param_attr = paddle.attr.Param(
        name='fc_param_attr_' + fc_param,
    )
    fc_bias_attr = paddle.attr.Param(
        name='fc_bias_attr_' + fc_param,
    )

    word_embedding = paddle.layer.embedding(
        input=seq_data,
        size=word_vector_dim,
        name=scope_name + '_WordEmbedding',
        param_attr=embedding_param,
    )
    word_sum = paddle.layer.pooling(
        input=word_embedding,
        pooling_type=paddle.pooling.Sum(),
        agg_level=paddle.layer.AggregateLevel.TO_NO_SEQUENCE
    )
    fc = paddle.layer.fc(
        input=word_sum,
        size=fc_dim,
        param_attr=fc_param_attr,
        bias_attr=fc_bias_attr,
    )
    return fc


def get_rnn_model(scope_name, seq_data, word_vector_dim, rnn_dim, rnn_scope):
    embedding_param = paddle.attr.Param(
        name='embedding_param',
        initial_max=1.0,
        initial_min=-1.0,
        initial_std=0.01,
    )
    mat_param = paddle.attr.Param(
        name='lstm_mat_param_' + rnn_scope,
    )
    bias_param = paddle.attr.Param(
        name='lstm_bias_param_' + rnn_scope,
    )
    inner_param = paddle.attr.Param(
        name='lstm_inner_param_' + rnn_scope,
    )

    gru_param_attr = paddle.attr.Param(
        name='gru_param_attr_' + rnn_scope,
    )
    gru_bias_attr = paddle.attr.Param(
        name='gru_bias_attr_' + rnn_scope,
    )

    word_embedding = paddle.layer.embedding(
        input=seq_data,
        size=word_vector_dim,
        name=scope_name + '_WordEmbedding',
        param_attr=embedding_param,
    )
    sentence_forward = paddle.networks.simple_lstm(
        input=word_embedding,
        size=rnn_dim,
        name=scope_name + '_ForwardRnn',
        mat_param_attr=mat_param,
        bias_param_attr=bias_param,
        inner_param_attr=inner_param,
    )
    # sentence_forward = paddle.networks.simple_gru(
    #     input=word_embedding,
    #     size=rnn_dim,
    #     name=scope_name + '_ForwardRnn',
    #     gru_param_attr=gru_param_attr,
    #     gru_bias_attr=gru_bias_attr,
    # )
    last_sentence = paddle.layer.last_seq(input=sentence_forward)
    return last_sentence


def get_sum_rnn_model(scope_name, seq_data, word_vector_dim, rnn_dim, rnn_scope):
    embedding_param = paddle.attr.Param(
        name='embedding_param',
        initial_max=1.0,
        initial_min=-1.0,
        initial_std=0.01,
    )
    mat_param = paddle.attr.Param(
        name='lstm_mat_param_' + rnn_scope,
    )
    bias_param = paddle.attr.Param(
        name='lstm_bias_param_' + rnn_scope,
    )
    inner_param = paddle.attr.Param(
        name='lstm_inner_param_' + rnn_scope,
    )
    word_embedding = paddle.layer.embedding(
        input=seq_data,
        size=word_vector_dim,
        name=scope_name + '_WordEmbedding',
        param_attr=embedding_param,
    )
    word_sum = paddle.layer.pooling(
        input=word_embedding,
        pooling_type=paddle.pooling.Sum(),
        agg_level=paddle.layer.AggregateLevel.TO_SEQUENCE
    )
    sentence_forward = paddle.networks.simple_lstm(
        input=word_sum,
        size=rnn_dim,
        name=scope_name + '_ForwardRnn',
        mat_param_attr=mat_param,
        bias_param_attr=bias_param,
        inner_param_attr=inner_param,
    )
    return sentence_forward


def build_rnn_model(dict_size):
    word_vector_dim = 64
    rnn_dim = 32

    # encoder
    query_seq = paddle.layer.data(
        name='query_seq',
        type=paddle.data_type.integer_value_sequence(dict_size),
    )
    title_1_seq = paddle.layer.data(
        name='title_1_seq',
        type=paddle.data_type.integer_value_sequence(dict_size),
    )
    title_2_seq = paddle.layer.data(
        name='title_2_seq',
        type=paddle.data_type.integer_value_sequence(dict_size),
    )
    label_data = paddle.layer.data(
        name='label',
        type=paddle.data_type.integer_value(1)
    )
    get_basic_model = get_sum_model
    # get_basic_model = get_rnn_model
    # get_basic_model = get_sum_rnn_model
    query_vector = get_basic_model('query', query_seq, word_vector_dim, rnn_dim, 'query')
    title_1_vector = get_basic_model('title_1', title_1_seq, word_vector_dim, rnn_dim, 'title')
    title_2_vector = get_basic_model('title_2', title_2_seq, word_vector_dim, rnn_dim, 'title')
    sim_1 = paddle.layer.cos_sim(a=query_vector, b=title_1_vector, name='cosine_1')
    sim_2 = paddle.layer.cos_sim(a=query_vector, b=title_2_vector, name='cosine_2')
    # delta_sim=sim_1-sim_2
    # evaluator = paddle.evaluator.auc(input=sim_1, label=label_data)
    cost = paddle.layer.rank_cost(
        left=sim_1,
        right=sim_2,
        label=label_data,
        # evaluator=paddle_helper.pnpair_evaluator()
    )
    return cost


def cluster_main():
    paddle.init(
        use_gpu=use_gpu,
        trainer_count=int(os.getenv("PADDLE_TRAINER_COUNT", "1")),
        port=int(os.getenv("PADDLE_PORT", "7164")),
        ports_num=int(os.getenv("PADDLE_PORTS_NUM", "1")),
        ports_num_for_sparse=int(os.getenv("PADDLE_PORTS_NUM_FOR_SPARSE", "1")),
        num_gradient_servers=int(os.getenv("PADDLE_NUM_GRADIENT_SERVERS", "1")),
        trainer_id=int(os.getenv("PADDLE_TRAINER_ID", "0")),
        pservers=os.getenv("PADDLE_PSERVERS", "127.0.0.1"))

    is_train = True
    dict_size = 1000000
    feeding = {
        'query_seq': 0,
        'title_1_seq': 1,
        'title_2_seq': 2,
        'label': 3,
    }

    cost = build_rnn_model(dict_size)
    if is_train:
        train_batch_reader = paddle.batch(
            paddle.reader.shuffle(
                # reader_pairwise_content(
                reader_pairwise_title(
                    data_filedir=cluster_train_dir,
                    dict_filename='thirdparty/prm_thirdparty/word_list.txt',
                    dict_size=dict_size,
                ),
                buf_size=1000,
            ),
            batch_size=200,
        )
        test_batch_reader = paddle.batch(
            paddle.reader.shuffle(
                # reader_pairwise_content(
                reader_pairwise_title(
                    data_filedir=cluster_test_dir,
                    dict_filename='thirdparty/prm_thirdparty/word_list.txt',
                    dict_size=dict_size,
                ),
                buf_size=1000,
            ),
            batch_size=200,
        )
        # define optimize method and trainer
        optimizer = paddle.optimizer.Adam(
            learning_rate=1e-4,
            regularization=paddle.optimizer.L2Regularization(rate=1e-3))
        parameters = paddle.parameters.create(cost)
        trainer = paddle.trainer.SGD(
            cost=cost, parameters=parameters, update_equation=optimizer,
            is_local=False,
        )

        # define event_handler callback
        def event_handler(event):
            if isinstance(event, paddle.event.EndIteration):
                if event.batch_id % 10 == 0:
                    train_info = "\nPass %d, Batch %d, Cost %f, %s" % \
                                 (event.pass_id, event.batch_id, event.cost, event.metrics)
                    print(train_info)
                    if not os.path.exists("output"):
                        os.makedirs("output")
                    with open('output/train_info', 'w+') as f:
                        f.write(train_info)
                        # print event.gm.getLayerOutputs("cosine_1")
                        # print event.gm.getLayerOutputs("cosine_2")
                        # else:
                        #     sys.stdout.write('.')
                        #     sys.stdout.flush()

                        # if not event.batch_id % 10:
                        #     save_path = 'params_pass_%05d_batch_%05d.tar' % (
                        #         event.pass_id, event.batch_id)
                        #     save_model(parameters, save_path)

            if isinstance(event, paddle.event.EndPass):
                result = trainer.test(reader=test_batch_reader, feeding=feeding)
                test_info = "\nTest with Pass %d, Cost %f, %s" % (event.pass_id, result.cost, result.metrics)
                print(test_info)
                if not os.path.exists("output"):
                    os.makedirs("output")
                with open('output/test_info', 'w+') as f:
                    f.write(test_info)
                # save parameters
                # save_path = 'params_pass_%05d.tar' % (event.pass_id)
                # save_model(parameters, save_path)
                if not os.path.exists("output"):
                    os.makedirs("output")
                model_file = "output/dnn_params_pass" + str(event.pass_id) + ".tar.gz"
                with gzip.open(model_file, 'w') as f:
                    parameters.to_tar(f)

        trainer.train(
            reader=train_batch_reader,
            feeding=feeding,
            event_handler=event_handler,
            num_passes=10,
        )
    else:
        model_path = 'thirdparty/prm_thirdparty/dnn_params_test.tar.gz'
        parameters = paddle.parameters.Parameters.from_tar(gzip.open(model_path))


if __name__ == '__main__':
    # 待定：sum or rnn; title or content
    cluster_main()

```

**Comment by prm10** [comment]
设置is_local=false之后，训练集与测试集相同时，测试集的cost会下降了。
两者数据不同时，测试集cost仍然不下降，应该是我的模型或者测试集数据有问题了。
paddle集群的问题应该没有了，多谢~
